internal/validator/default/helpers.go

```diff
@@ -56,110 +57,71 @@ func extendTx(ctx context.Context, f validator.TxFinderI, rawTx *sdkTx.Transacti
-			return err
+	return nil
+}
+
+func extendInputs(tx *sdkTx.Transaction, childInputs []*sdkTx.TransactionInput) error {
+	for _, input := range childInputs {
+		if len(tx.Outputs) < int(input.SourceTxOutIndex) {
+			return fmt.Errorf("output %d not found in transaction %s", input.SourceTxOutIndex, input.PreviousTxIDStr())
 		}
+		output := tx.Outputs[input.SourceTxOutIndex]
+
+		input.SetPrevTxFromOutput(output)
 	}
 
 	return nil
 }
 
-// getUnminedAncestors returns unmined ancestors with data necessary to perform Deep Fee validation
-func getUnminedAncestors(ctx context.Context, w validator.TxFinderI, tx *sdkTx.Transaction, tracingEnabled bool, tracingAttributes ...attribute.KeyValue) (map[string]*sdkTx.Transaction, error) {
+// getUnminedAncestors returns unmined ancestors with data necessary to perform cumulative fee validation
+func getUnminedAncestors(ctx context.Context, txFinder validator.TxFinderI, tx *sdkTx.Transaction, tracingEnabled bool, tracingAttributes ...attribute.KeyValue) (map[string]*sdkTx.Transaction, error) {
 	ctx, span := tracing.StartTracing(ctx, "getUnminedAncestors", tracingEnabled, tracingAttributes...)
 	defer tracing.EndTracing(span)
 	unmindedAncestorsSet := make(map[string]*sdkTx.Transaction)
 
 	// get distinct parents
 	// map parentID with inputs collection to avoid duplication and simplify later processing
-	parentInputMap := make(map[string][]*sdkTx.TransactionInput)
+	parentInputMap := make(map[string]struct{})
 	parentsIDs := make([]string, 0, len(tx.Inputs))
 
 	for _, in := range tx.Inputs {
 		prevTxID := in.PreviousTxIDStr()
-
-		inputs, found := parentInputMap[prevTxID]
+		_, found := parentInputMap[prevTxID]
 		if !found {
 			// first occurrence of the parent
-			inputs = make([]*sdkTx.TransactionInput, 0)
 			parentsIDs = append(parentsIDs, prevTxID)
 		}
-
-		inputs = append(inputs, in)
-		parentInputMap[prevTxID] = inputs
 	}
-
-	// get parents
-	const finderSource = validator.SourceTransactionHandler | validator.SourceWoC
-	parentsTxs, err := w.GetRawTxs(ctx, finderSource, parentsIDs)
+	mempoolAncestorTxIDs, err := txFinder.GetMempoolAncestors(ctx, parentsIDs)
 	if err != nil {
-		return nil, errors.Join(ErrFailedToGetRawTxs, fmt.Errorf("parent: %v", parentsIDs), err)
+		return nil, errors.Join(ErrFailedToGetMempoolAncestors, err)
 	}
 
-	if len(parentsTxs) != len(parentsIDs) {
-		return nil, ErrParentNotFound
-	}
-
-	for _, p := range parentsTxs {
-		if _, found := unmindedAncestorsSet[p.TxID]; found {
-			continue // parent was proccesed already
-		}
+	allTxIDs := append(parentsIDs, mempoolAncestorTxIDs...)
 
-		childInputs, found := parentInputMap[p.TxID]
-		if !found {
-			return nil, ErrParentNotFound
-		}
-
-		// fulfill data about the parent for further validation
-		bTx, err := sdkTx.NewTransactionFromBytes(p.Bytes)
-		if err != nil {
-			return nil, fmt.Errorf("cannot parse parent tx: %w", err)
-		}
+	ancestorTxs, err := txFinder.GetRawTxs(ctx, validator.SourceNodes, allTxIDs)
```

Comment:
Shouldn't be `validator.SourceTransactionHandler | validator.SourceWoC` ?

Reply:
Yes, but I think we can actually put here again all sources

```go
const finderSource = validator.SourceTransactionHandler | validator.SourceNodes | validator.SourceWoC
```